code/obj/submachine/seed.dm

```diff
@@ -1,632 +1,491 @@
+					thisContainerData["finalColor"] = R.get_average_rgb()
+					// Reagent data
+					for(var/reagent_id in R.reagent_list)
+						var/datum/reagent/current_reagent = R.reagent_list[reagent_id]
 
-			else
-				dat += {"<b>Splice Error.</b><br>
-				<A href='?src=\ref[src];page=3'>Please click here to return to the Seed List.</A>"}
-		else
-			dat += {"<b>Software Error.</b><br>
-			<A href='?src=\ref[src];page=1'>Please click here to return to the Overview.</A>"}
+						contents.Add(list(list(
+							name = reagents_cache[reagent_id],
+							id = reagent_id,
+							colorR = current_reagent.fluid_r,
+							colorG = current_reagent.fluid_g,
+							colorB = current_reagent.fluid_b,
+							volume = current_reagent.volume
+						)))
 
-		dat += {"<hr>
-		Genetic display key: <span class='hyp-dominant'>Dominant</span> / Recessive
-		"}
+		if(src.splicing1)
+			splice1_geneout = QuickAnalysisRow(src.splicing1, src.splicing1.planttype, src.splicing1.plantgenes)
+			splice1_geneout["damage"] = list(src.splicing1.seeddamage, FALSE)
+			splice1_geneout["splicing"] = list("splicing", TRUE)
+			splice1_geneout["allow_infusion"]= list("allow_infusion", src.inserted?.reagents?.total_volume > 0)
+		if(src.splicing2)
+			splice2_geneout = QuickAnalysisRow(src.splicing2, src.splicing2.planttype, src.splicing2.plantgenes)
+			splice2_geneout["damage"] = list(src.splicing2.seeddamage, FALSE)
+			splice2_geneout["splicing"] = list("splicing", TRUE)
+			splice2_geneout["allow_infusion"]= list("allow_infusion", src.inserted?.reagents?.total_volume > 0)
+
+		return list(\
+			"extractables" = exlist,\
+			"seeds" = seedlist,\
+			"category" = src.mode,\
+			"category_lengths" = list(length(src.extractables),length(src.seeds)),\
+			"inserted" =  src.inserted ? "[src.inserted.reagents.total_volume]/[src.inserted.reagents.maximum_volume] [src.inserted.name]" : "No reagent vessel",\
+			"inserted_container" = thisContainerData,\
+			"seedoutput" = src.seedoutput,\
+			"splice_chance" = splice_chance,\
+			"show_splicing" = src.splicing1 || src.splicing2,
+			"splice_seeds" = list(splice1_geneout, splice2_geneout),
+			"sortBy" = src.sort,
+			"sortAsc" = src.sortAsc,
+		)
 
-		user.Browse(jointext(dat, ""), "window=plantmaster;size=800x400")
-		onclose(user, "rextractor")
+	ui_interact(mob/user, datum/tgui/ui)
+		ui = tgui_process.try_update_ui(user, src, ui)
+		if(!ui)
+			ui = new(user, src, "Plantmaster")
+			ui.open()
 
-	Topic(href, href_list)
-		if((BOUNDS_DIST(usr, src) > 0) && !issilicon(usr) && !isAI(usr))
-			boutput(usr, "<span class='alert'>You need to be closer to the machine to do that!</span>")
-			return
-		if(href_list["page"])
-			var/ops = text2num_safe(href_list["page"])
-			switch(ops)
-				if(2) src.mode = "extraction"
-				if(3) src.mode = "seedlist"
-				else src.mode = "overview"
-			playsound(src.loc, "sound/machines/click.ogg", 50, 1)
-			src.updateUsrDialog()
+	ui_act(action, list/params, datum/tgui/ui)
```

Comment:
Missing a lil code here that would otherwise allow you to interact with this machine, even if not in range for example

```suggestion
	ui_act(action, list/params, datum/tgui/ui)
		. = ..()
		if(.)
			return
```